src/mds/Server.cc

```diff
@@ -8049,11 +8059,33 @@ void Server::_unlink_local(MDRequestRef& mdr, CDentry *dn, CDentry *straydn)
     // will manually pop projected inode
 
     // primary link.  add stray dentry.
-    mdcache->predirty_journal_parents(mdr, &le->metablob, in, dn->get_dir(), PREDIRTY_PRIMARY|PREDIRTY_DIR, -1);
-    mdcache->predirty_journal_parents(mdr, &le->metablob, in, straydn->get_dir(), PREDIRTY_PRIMARY|PREDIRTY_DIR, 1);
+    if (dnl->is_primary()) {
+      mdcache->predirty_journal_parents(mdr, &le->metablob, in, dn->get_dir(), PREDIRTY_PRIMARY|PREDIRTY_DIR, -1);
+      mdcache->predirty_journal_parents(mdr, &le->metablob, in, straydn->get_dir(), PREDIRTY_PRIMARY|PREDIRTY_DIR, 1);
 
-    pi.inode->update_backtrace();
-    le->metablob.add_primary_dentry(straydn, in, true, true);
+      pi.inode->update_backtrace();
+      le->metablob.add_primary_dentry(straydn, in, true, true);
+    } else {
+      // referent remote link - to purge the referent inode created
+      auto pri = ref_in->project_inode(mdr);
+      {
+        std::string t;
+        dn->make_path_string(t, true);
+        pri.inode->stray_prior_path = std::move(t);
+      }
+      pri.inode->version = ref_in->pre_dirty();
+      // Purge enqueue requires inode to be primary and nlink to be 0
+      pri.inode->nlink--;
+      if (pri.inode->nlink == 0)
+        ref_in->state_set(CInode::STATE_ORPHAN);
+      mdcache->predirty_journal_parents(mdr, &le->metablob, ref_in, straydn->get_dir(), PREDIRTY_PRIMARY|PREDIRTY_DIR, 1);
+      pri.inode->update_backtrace();
+      le->metablob.add_primary_dentry(straydn, ref_in, true, true);
+
+      mdcache->predirty_journal_parents(mdr, &le->metablob, in, dn->get_dir(), PREDIRTY_DIR, -1);
+      mdcache->predirty_journal_parents(mdr, &le->metablob, in, 0, PREDIRTY_PRIMARY);
```

Comment:
please use `nullptr` for NULL pointer types - in this case for the parent dir pointer